core/src/main/java/io/aiven/kafka/tieredstorage/metadata/SegmentCustomMetadataBuilder.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.metadata;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+
+import io.aiven.kafka.tieredstorage.ObjectKey;
+
+public class SegmentCustomMetadataBuilder {
+    final ObjectKey objectKey;
+    final RemoteLogSegmentMetadata segmentMetadata;
+    final Map<ObjectKey.Suffix, UploadResult> uploadResults;
+
+    final Set<SegmentCustomMetadataField> fields;
+
+    public SegmentCustomMetadataBuilder(final Set<SegmentCustomMetadataField> fields,
+                                        final ObjectKey objectKey,
+                                        final RemoteLogSegmentMetadata segmentMetadata) {
+        this.fields = fields;
+        this.objectKey = objectKey;
+        this.segmentMetadata = segmentMetadata;
+        this.uploadResults = new HashMap<>(ObjectKey.Suffix.values().length);
+    }
+
+    public SegmentCustomMetadataBuilder addUploadResult(final ObjectKey.Suffix suffix,
+                                                        final String keyPath,
+                                                        final long bytes) {
+        if (uploadResults.containsKey(suffix)) {
+            throw new IllegalArgumentException("Upload results for suffix " + suffix + " already added");
+        }
+        uploadResults.put(suffix, new UploadResult(keyPath, bytes));
+        return this;
+    }
+
+    public long totalSize() {
+        return uploadResults.values().stream().mapToLong(value -> value.bytes).sum();
+    }
+
+    public NavigableMap<Integer, Object> build() {
+        if (fields.isEmpty()) {
+            return new TreeMap<>();
+        }
+        final TreeMap<Integer, Object> taggedFields = new TreeMap<>();
+        fields.forEach(field -> taggedFields.put(field.ordinal(), field.valueProvider.apply(this)));
+        return taggedFields;
+    }
+
+    private static class UploadResult {
```

Comment:
Can you elaborate why do we need this class if we are not using it externally?

Reply:
Was initially planning to use it for other metadata fields including paths. But as you spotted, no need for it, and Long is enough.